survivalstan/formulas.py

```diff
@@ -0,0 +1,414 @@
+import pandas as pd
+import patsy
+import sys
+import numpy as np
+import re
+import logging
+logger = logging.getLogger(__name__)
+
+def _prep_timepoint_dataframe(df,
+                              timepoint_end_col,
+                              timepoint_id_col = None
+                              ):
+    """ Helper function to take a set of timepoints
+        in observation-level dataframe & return
+        formatted timepoint_id, end_time, duration
+
+        Returns
+        ---------
+        pandas dataframe with one record per timepoint_id
+            where timepoint_id is the index
+            sorted on the index, increasing
+
+    """
+    time_df = df.copy()
+    time_df.sort_values(timepoint_end_col, inplace=True)
+    if not(timepoint_id_col):
+        timepoint_id_col = 'timepoint_id'
```

Comment:
Not really - in the current design, the actions to be performed are a function of the set of parameters given.

E.g. if a user doesn't supply either timepoint_id or timepoint_id_col, then no timepoint id col is generated. However if there is a timepoint_id but no timepoint_id_col, then we want to use a default name for timepoint_id_col.

One thing I want to move away from (and one motivation for the `surv` syntax) is the current method of having paired *_id / *_id_col names for every parameter. This is the rational for the 'id' construct, which retains its state. 

So the current *_id & *_id_col is remaining in part for backwards compatibility, but the _id_col part of it isn't strictly necessary any longer. Prob better to chat offline, about one or both of these: (1) design confusion, and (2) is it worth keeping *_id and *_id_col pairs hanging around.